### ISSUE TITLE
Deploy TREEHOOSE

### ISSUE THREAD

**Issue by jemrobinson** [alan-turing-institute/data-safe-haven-team#2]
We would like to evaluate [TREEHOOSE](https://github.com/HicResearch/TREEHOOSE) and compare its features to the Data Safe Haven.

- ~~Needs a submission to RCP team to ask for AWS credits~~ See @edwardchalstrey1 comment below
- [x] Consider talking to Dundee/Sheffield/UCL/Liverpool for possible collaboration/deployment advice
  - Dundee: Simon Li
  - Sheffield: Lucy Cheesman
  - UCL: Nel Swanepoel
  - Liverpool: Gary Leeming and David Salac

## What are we looking for?

See #5 for details

## Cost

see #6

**Comment by jemrobinson** [comment]
## 2022-10-26 Discussion with Simon and Ed 

- Need a new AWS account (similar to an Azure subscription) for every project
    - has a root username and password
    - use this to setup IAM users and groups
- Using an AWS organisation (similar to an Azure tenancy) allows SSO across multiple accounts
    - may want to consider setting up an organisation but the deployment might work without one
- AWS Control Tower to impose standards across accounts
- AppStream VDI for remote desktop
    - similar to eg. Guacamole/RDS
    - needs to be manually configured
- User management for admins through browser interface
    - AWS Cognito for login
- User identity **not** bound to workspace
- Service Workbench is a separate codebase for deploying a thing that could be a component of a TRE but isn't one on its own
    - Amazon own the service workbench repo (awslabs/service-workbench-on-aws)
- TREEHOOSE is currently one TRE per project
- Pasting is allowed (rationale: this is the same as running a virtual keyboard)
- Only Windows VMs supported for compute

**Comment by edwardchalstrey1** [comment]
### Setting up AWS at Turing

Steps that were taken by the Trustworthy ID project:

1. price it up with https://calculator.aws/#/
2. get PMU to authorise the budget
3. then get IT to set up the account - which in our case we'll probably do ourselves instead of relying on IT

**Comment by jemrobinson** [comment]
Presumably someone needs to add a credit card or set up invoicing? Can you check with the Gates PMU contact how that was done? For us I think @Arielle-Bennett is the right person to talk to.

**Comment by edwardchalstrey1** [comment]
> Presumably someone needs to add a credit card or set up invoicing? Can you check with the Gates PMU contact how that was done? For us I think @Arielle-Bennett is the right person to talk to.

Ok, I'm going to send an email now - in the meantime while we wait for a response, I guess we can at some point look at the AWS calculator and plug in the anticipated components needed for Treehoose

**Comment by manics** [comment]
I've opened a TREEHOOSE PR to update the version of the underlying Service Workbench, along with some minor docs improvements
https://github.com/HicResearch/TREEHOOSE/pull/89
I've got another repo with a GitHub deployment workflow which I can anonymise and open-source if it's useful.

**Comment by edwardchalstrey1** [comment]
Hi @manics it would be handy if you could comment on #6 - we're looking to get a rough estimate before we request budget for our TREEHOOSE deployment

**Comment by edwardchalstrey1** [comment]
@jemrobinson to keep to appraised as to progress: At the moment, RCP are blocked by having to deal with questions from AWS support - @fedenanni suggested it could be ready before the start of December, but even that we aren't 100% sure

**Comment by fedenanni** [comment]
@jemrobinson @edwardchalstrey1 I'd keep my expectation low on "the start of December" because AWS seems quite slow in processing our change in the way payments are handled (from credit card to invoice), which is a prerequisite before setting up your account. If there's any progress during the week I'll let you know, otherwise I think it would be safe to just postpone everything to January as it seems that most of us will be on leave during December.

**Comment by edwardchalstrey1** [comment]
@fedenanni are RCP still blocked with progress on the Turing AWS "account" (or organisation, whatever the correct term is)?

**Comment by edwardchalstrey1** [comment]
Hi @manics I'm yet to attempt a deployment of TREEHOOSE for the purpose of comparing and contrasting with Turing-DSH and Azure TRE, but in the meantime I wondered if there was any user documentation for researchers working in a TREEHOOSE TRE, or alternatively some part of the developer/operations docs that explains what a researcher would see when logged into a desktop of a TRE instance, which apps are available etc.

The [github docs](https://github.com/HicResearch/TREEHOOSE#documentation) mostly seem to pertain to the AWS infrastructure and operations etc as oppose to the features of TREEHOOSE. For example in DSH we have a [list of available software](https://alan-turing-institute.github.io/data-safe-haven/v4.0.1/roles/researcher/available_software.html), and the user guide shows researchers how to use things like GitLab, CodiMD alongside the usual account creation and setup steps

**Comment by manics** [comment]
There's a few screenshots in https://www.manicstreetpreacher.co.uk/hic-presentations-public/20220329-ukri-cloud-workshop-talk/
We're working on a proper website for our user-facing docs, but there's some info in
https://hic-docs.atlassian.net/wiki/spaces/HKB/pages/870580396/How-to+articles

The available applications are down to the TRE administrator- we build all our machine images with Packer but by default there are no applications in the workspaces- open-sourcing the repo with our Packer templates is on our todo list.

**Comment by edwardchalstrey1** [comment]
Thanks Simon!

**Comment by edwardchalstrey1** [comment]
# Deployment logs

[Step 1A of prerequisites](https://github.com/alan-turing-institute/TREEHOOSE/blob/main/doc/deployment/Prerequisites.md#step-1a-create-encryption-key-for-aws-control-tower) I am getting a permissions issue error when attempting to upload the yaml:
Screenshot 2023-01-10 at 11.29.10

**Comment by manics** [comment]
Skip all the landing-zone stuff, that needs to be done in the top-level organisation/management account. The idea is your organisation admin can set policies that apply to all AWS accounts in the organisation. Since you've only got a single account you won't have access to it.

**Comment by manics** [comment]
Having said that.... this will block other cloudformation templates. Since you're using an SSO account you'll assume a role chosen by the org admin (you might have a choice of roles), can you ask them for all the permissions they've granted to you?

**Comment by edwardchalstrey1** [comment]
Turing IT have advised that I am in a group that has [SystemAdministrator privileges](https://docs.aws.amazon.com/IAM/latest/UserGuide/access_policies_job-functions.html#jf_system-administrator), and that "We have top-level polices that deny creating IAM accounts, deleting CloudWatch logs and leaving the organization."
Makes sense because when I log into management console I did so as SystemAdministrator:
![Screenshot 2023-01-10 at 14 03 15](https://user-images.githubusercontent.com/5486164/211572492-200c77ab-60e2-4bdc-bd07-adc101dc383e.png)

**Comment by manics** [comment]
https://github.com/z0ph/MAMIP/blob/master/policies/SystemAdministrator
CloudFormation is notable for it's absence!

https://github.com/HicResearch/TREEHOOSE/blob/7d4faa178bda968dc8c818a636ae0eea89f6b5f3/src/deployment/DeploymentInstance-Cfn.yaml#L262-L541
from
https://github.com/HicResearch/TREEHOOSE/pull/89/

contains the latest minimal set of permissions I could find, though it might still assume the presence of some very basic setup (like the ability to run cloudformation in the first place).

**Comment by edwardchalstrey1** [comment]
Ok, sounds like that could be problematic then! I will discuss internally how feasible getting higher privileges is

**Comment by edwardchalstrey1** [comment]
After chatting with @manics we concluded the most important thing in the pre-requisites was [AppStream](https://github.com/manics/TREEHOOSE/blob/update-5.2.3/doc/deployment/Prerequisites.md#step-4-initialise-appstream), however looks like even this will require the `cloudformation:DescribeStacks` action

**Comment by edwardchalstrey1** [comment]
Hi @manics unfortunately we're still struggling to get an AWS setup with the correct privileges to deploy Treehoose here at Turing - I asked this before but unfortunately I don't remember your answer - as an alternative, or at least in the meantime, is there a Sandbox Treehoose environment you have already deployed that I could be given user access to?

**Comment by manics** [comment]
@edwardchalstrey1 Sure! Are you OK with waiting a few days so I can update the deployment (it's a bit behind at the moment)?

**Comment by edwardchalstrey1** [comment]
> @edwardchalstrey1 Sure! Are you OK with waiting a few days so I can update the deployment (it's a bit behind at the moment)?

Absolutely, no rush on this, any help at all is a bonus!

**Comment by edwardchalstrey1** [comment]
Closing this since this is not something we're currently still planning to do - instead, Simon has provided me with a login to https://treehoose.dev.hic.dundee.ac.uk/ so we can use this for comparison work
